src/core/overflowsafe_type.hpp

```diff
@@ -134,72 +150,81 @@ class OverflowSafeInt
+	inline OverflowSafeInt  operator /  (const intmax_t         divisor) const { return OverflowSafeInt(static_cast<intmax_t>(this->m_value) / static_cast<intmax_t>(divisor)); }
+	inline OverflowSafeInt& operator /= (const OverflowSafeInt& divisor)       { return *this /= divisor.m_value; }
+	inline OverflowSafeInt  operator /  (const OverflowSafeInt& divisor) const { return *this /  divisor.m_value; }
 
 	/* Operators for modulo */
-	inline OverflowSafeInt& operator %= (const int  divisor)       { this->m_value %= divisor; return *this; }
-	inline OverflowSafeInt  operator %  (const int  divisor) const { OverflowSafeInt result = *this; result %= divisor; return result; }
+	template <class I, class = typename std::enable_if<std::is_integral<I>::value && (static_cast<intmax_t>(std::numeric_limits<I>::min()) >= static_cast<intmax_t>(T_MIN)) && (static_cast<uintmax_t>(std::numeric_limits<I>::max()) <= static_cast<uintmax_t>(T_MAX))>::type>
+	inline OverflowSafeInt& operator %= (const I                divisor)       { this->m_value %= static_cast<T>(divisor); return *this; }
+	template <class I, class = typename std::enable_if<std::is_integral<I>::value && (static_cast<intmax_t>(std::numeric_limits<I>::min()) >= static_cast<intmax_t>(T_MIN)) && (static_cast<uintmax_t>(std::numeric_limits<I>::max()) <= static_cast<uintmax_t>(T_MAX))>::type>
+	inline OverflowSafeInt  operator %  (const I                divisor) const { OverflowSafeInt result = *this; result %= static_cast<T>(divisor); return result; }
+	inline OverflowSafeInt& operator %= (const intmax_t         divisor)       { *this = *this % divisor; return *this; }
+	inline OverflowSafeInt  operator %  (const intmax_t         divisor) const { return OverflowSafeInt(static_cast<intmax_t>(this->m_value) % static_cast<intmax_t>(divisor)); }
+	inline OverflowSafeInt& operator %= (const OverflowSafeInt& divisor)       { return *this %= divisor.m_value; }
+	inline OverflowSafeInt  operator %  (const OverflowSafeInt& divisor) const { return *this %  divisor.m_value; }
 
 	/* Operators for shifting */
-	inline OverflowSafeInt& operator <<= (const int shift)       { this->m_value <<= shift; return *this; }
-	inline OverflowSafeInt  operator <<  (const int shift) const { OverflowSafeInt result = *this; result <<= shift; return result; }
-	inline OverflowSafeInt& operator >>= (const int shift)       { this->m_value >>= shift; return *this; }
-	inline OverflowSafeInt  operator >>  (const int shift) const { OverflowSafeInt result = *this; result >>= shift; return result; }
+	inline OverflowSafeInt  operator <<  (const int shift) const { return OverflowSafeInt(this->m_value << shift); }
+	inline OverflowSafeInt& operator <<= (const int shift)       { *this = *this << shift; return *this; }
+	inline OverflowSafeInt  operator >>  (const int shift) const { return OverflowSafeInt(this->m_value >> shift); }
+	inline OverflowSafeInt& operator >>= (const int shift)       { *this = *this >> shift; return *this; }
```

Comment:
Typecast is supposed to ensure that the value is not shifted outside the valid range; however, I just realised this won't actually work :wink:.